src/Engine/Plugins/EngineControlPlugin.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <thread>
+#include <queue>
+#include <functional>
+#include <memory>
+
+#include "Platform/Proxy/ProxyOpenGLContext.h"
+#include "Platform/Proxy/ProxyEventLoop.h"
+
+#include "EngineControlStateHandle.h"
+
+class EngineController;
+class EngineControlState;
+class PlatformEvent;
+
+/**
+ * This plugin exposes a coroutine-like API that makes it possible to control the game by passing in synthetic
+ * platform events.
+ *
+ * The implementation just uses a separate thread that runs a control routine, and the execution can switch between
+ * threads from inside the `SwapBuffers` call. This effectively means that the control routine runs in between game
```

Comment:
That's a discussion we should have when we plan our multi threading approach. But I doubt we would ever need it tbh.

Reply:
well all different updates types CAN be implemented in different threads, but that's not necessary ..

Reply:
Tbh I didn't yet think that far into the future. Need to clean up the codebase first.

The reason I hooked into SwapBuffers is very simple --- this was the only solution that was within my reach that didn't involve rewriting half the engine. And I cannot rewrite half the engine, because we don't have tests. And to have tests we need tracing X).

I do realize that I'll have to do some major changes in the tracing system eventually.